KW41 PWM Support

alexandru-porosanu-nxp opened the pull request:
This patch-set adds support for the TPM driver for KW4x family of SoCs. The TPM driver is based on the RV32M1 TPM driver, and the corresponding SDK driver.
This patch-set depends on https://github.com/zephyrproject-rtos/hal_nxp/pull/39

alexandru-porosanu-nxp committed:
drivers: pwm: add driver for NXP Kinetis TPM module

The TPM (Timer/PWM Module) is a 2- to 8-channel timer which supports
input capture, output compare, and the generation of PWM signals to
control electric motor and power management applications.

This patch adds the driver and the binding necessary for instantiating
the driver. The work is based on the RV32M1 driver for TPM done by
Henrik Brix Andersen. A later patch will enable this driver to be used
for the KW41Z SoC, if PWM support is requested.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

alexandru-porosanu-nxp committed:
soc: kinetis: add support for TPM PWM module

Some Kinetis SoCs have an instance of the the TPM module
that can be used for PWM control. As such, add the necessary
configurations to enable it on the SoCs that support it, as well as
enable the clock for the module to function.

In this case, the enablement is done only for the KW41Z SoCs,
but there are other SoCs that support it, f.i. KW38Z

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

alexandru-porosanu-nxp committed:
boards: frdm_kw41z: enable PWM support

Once the TPM driver and the required configs have been added,
now this patch enables the usage of the PWM framework on KW41Z
SoCs. As such, the DTS gets the proper nodes and the pinmuxing
is done according to the configuration requested.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

alexandru-porosanu-nxp committed:
west.yml: hal_nxp: update HAL reference for MCUX TPM driver

Sync the NXP HAL commit in west with the one that enables
the TPM driver in HAL.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

galak commented on boards/arm/frdm_kw41z/pinmux.c:
```diff
@@ -28,9 +28,15 @@ static int frdm_kw41z_pinmux_init(struct device *dev)
 	/* Red, green, blue LEDs. Note the red LED and accel INT1 are both
 	 * wired to PTC1.
 	 */
+#if CONFIG_PWM_0
```
make this based on nodelabel instead.

alexandru-porosanu-nxp replied:
OK, I'm probably missing something here, but   if I drop the CONFIG_PWM_x altogether, then since the corresponding node is in the dts and has its status set as "okay" then the pimmux will always be enabled for PWM operation.  And that's not desirable. In the case of PWM0, its output pins are shared with GPIO pins that control the same RGB LED on the Freedomboard. The user should be allowed to select which one of the two methods for controlling the RGB led.

More generic speaking, just because something is in the DTS and is "okay", it doesn't mean that it should be enabled by the pinmux. More specifically, on the KW41, now SPI, I2C, ADC etc. are all enabled, since the corresponding nodes have the corresponding status set to "okay". If the user would like to build a simpler app, it'd need to modify the DTS and set the corresponding nodes to "disabled", as opposed to disabling options in the menuconfig.

So my thought is that the selection of the pinmux should be done based on the existing HW resources AND the user desired configuration.

Let me know your thoughts on this.

galak commented on soc/arm/nxp_kinetis/kwx/soc_kw4xz.c:
```diff
@@ -69,6 +70,10 @@ static ALWAYS_INLINE void clock_init(void)
 	CLOCK_SetLpuartClock(LPUART0SRC_OSCERCLK);
 #endif
+
+#if CONFIG_PWM_0 || CONFIG_PWM_1 || CONFIG_PWM_2
```
use nodelabel here.

alexandru-porosanu-nxp replied:
I think it's the same discussion as above: why enable the clock (that draws current) to an IP that is not used, since the user doesn't need/want it

MaureenHelm commented on west.yml:
```diff
@@ -92,7 +92,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 60fe8c0ae6e24d165cf42770b95c7df22b778386
+      revision: pull/39/head
```
Adding DNM label until the hal pr gets merged and this gets updated to the new SHA.

alexandru-porosanu-nxp replied:
Yep, once zephyrproject-rtos/hal_nxp#39 is merged, I'll change the SHA accordingly.

MaureenHelm replied:
It's been merged now

alexandru-porosanu-nxp replied:
> It's been merged now

Thanks @MaureenHelm , I've seen it and prepared the west.yml update. I'd need an input from @galak  , so I don't do two pushes, one with the west.yml changes and one with the potential DT flag changes

alexandru-porosanu-nxp commented:
I've redone the patchset, and incorporated the suggested nodelabel macros, but with an added condition that CONFIG_PWM is enabled.
Also, I've updated the west.yml to point to the appropriate SHA from NXP HAL

alexandru-porosanu-nxp commented:
Repushed due to checkpatch errors and code leftover

alexandru-porosanu-nxp commented on boards/arm/frdm_kw41z/Kconfig.defconfig:
```diff
@@ -50,4 +51,14 @@ choice ADC_MCUX_ADC16_VREF
+config PWM_2
+	default y
+
+endif # PWM
```
Right, I missed that when doing the dt lables update.

MaureenHelm commented:
Please rebase and fix the merge conflicts

alexandru-porosanu-nxp commented:
> Please rebase and fix the merge conflicts

All done, rebased on latest master and re-pushed. rgb_led is still working on the remote KW41, so all should be good

MaureenHelm commented:
> > Please rebase and fix the merge conflicts
> 
> All done, rebased on latest master and re-pushed. rgb_led is still working on the remote KW41, so all should be good

thanks!

@galak please revisit

galak commented on drivers/pwm/pwm_mcux_tpm.c:
```diff
@@ -0,0 +1,204 @@
+#define TPM_DEVICE(n) \
+	static const struct mcux_tpm_config mcux_tpm_config_##n = { \
+		.base =	(TPM_Type *) \
+			DT_INST_REG_ADDR(n), \
```
style nit, do all of these need to be split across 2 lines?

galak reviewed (approved):
Style nit, but lgtm.

alexandru-porosanu-nxp commented:
I've just noticed the checkpatch warnings: the vendor matching regex is broken by 5b10fac97efdb444b49f5e3c3ac16048ce8661c6 since the format for the vendor-prefixes file is different in Linux and Zephyr. @MaureenHelm , @galak , I've filled https://github.com/zephyrproject-rtos/zephyr/issues/25012 to track this . I'll probably submit a quickfix as well (revert the vendor match regexp to the old one)
